File: 0189-rotate-array/0189-rotate-array.java
class Solution {
    Deque<Integer> dq = new ArrayDeque<>();
    
    public void rotate(int[] nums, int k) {
        
        // 회전을 위해 deque자료구조 사용
        for (int num : nums) {
            dq.add(num);
        }

        // k번 반복
        for (int i = 0; i < k; i++) {
            int elem = dq.pollLast();
            dq.addFirst(elem);
        }
        
        // 기존 배열에 넘겨주기
        for (int i = 0; i < nums.length; i++) {
            nums[i] = dq.pollFirst();
        }
        
        
    }
}

File: 0189-rotate-array/0189-rotate-array.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
       
    
        d = dict()
        n = len(nums)
        
        # k번 이후에 위치할 index를 key, 그 값을 value로 하여 dict에 저장한다.
        for i in range(n): # O(N)
            d[(i + k) % n] = nums[i]
            
        
        for i in range(n): # O(N)
            nums[i] = d[i]